examples/ExactlyOnce/Program.cs

```diff
@@ -15,788 +15,988 @@
-
+        public TimeSpan LocalTransactionOperationTimeout { get; set; }
 
         /// <summary>
-        ///     The transactional id stem for the word count processor.
+        ///     Producer config.
         /// </summary>
-        const string TransactionalIdPrefix_Aggregate = "aggregator-transaction-id";
-
+        public ProducerConfig ProducerConfig { get; set; }
 
         /// <summary>
-        ///     Name of the consumer group used in the map phase when consuming
-        ///     from the Topic_InputLines topic.
+        ///     Timeout used for consume calls.
         /// </summary>
-        const string ConsumerGroup_MapWords = "map-words-consumer-group";
+        public int ConsumeTimeout { get; set; } = 100;
 
         /// <summary>
-        ///     Name of the consumer group used in the reduce phase when
-        ///     consuming from the Topic_Words topic.
+        ///     Max messages to be committed.
         /// </summary>
-        const string ConsumerGroup_Aggregate = "aggregate-consumer-group";
+        public int CommitMaxMessages { get; set; } = 1;
 
         /// <summary>
-        ///     Name of the consumer group used when loading count state
-        ///     from Topic_Counts.
+        ///     Max period between commits.
         /// </summary>
-        const string ConsumerGroup_LoadState = "load-state-consumer-group";
+        public TimeSpan CommitPeriod { get; set; } = TimeSpan.MaxValue;
 
+        /// <summary>
+        ///     Recreate the producer on fatal errors like 
+        ///     producer fenced and continue processing.
+        /// </summary>
+        public bool RecreateProducerOnFatalErrors { get; set; } = true;
 
-        class FasterState : IDisposable
+        /// <summary>
+        ///     Consumer configuration, don't set for a producer only
+        ///     processor.
+        /// </summary>
+        public ConsumerConfig ConsumerConfig
         {
-            public FasterState(Partition partition)
-            {
-                this.Path = System.IO.Path.Join(System.IO.Path.GetTempPath(), "partition_" + (int)partition);
-                this.ObjectLog = Devices.CreateLogDevice(this.Path + "_obj.log");
-                this.Log = Devices.CreateLogDevice(this.Path + ".log");
-                this.Store = new FasterKV<string, int>(
-                    size: 1L << 20, // 1M cache lines of 64 bytes each = 64MB hash table
-                    logSettings: new LogSettings { LogDevice = this.Log, ObjectLogDevice = this.ObjectLog }
-                );
-                var funcs = new SimpleFunctions<string, int>((a, b) => a + b); // function used for read-modify-write (RMW).
-                this.Session = this.Store.NewSession(funcs);
+            get
+            {
+                return consumerConfig;
             }
+            set
+            {
+                if (value != null && consumer == null)
+                {
+                    consumerConfig = value;
```

Comment:
doing this sort of thing is far too tricky/clever !!! 

the producer and consumer don't need to be class properties - they are only needed in the context of the RunLoop - create them there.

Reply:
the consumer can be put as a local variable, but the producer needs to be recreated and the instance should be passed back and forth to CommitTransaction and it cannot be returned in case it throws